helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java

```diff
@@ -190,42 +184,44 @@ private InputFields readInput(String content)
   private ClusterState readClusterStateAndValidateInput(String clusterId, InputFields inputFields)
       throws InvalidParameterException {
 
+    // One instance can only exist in one of the list in InstanceChange.
+    // Validate the intersection is empty.
+    for (String instance : inputFields.activatedInstances) {
+      if (inputFields.deactivatedInstances.contains(instance)) {
+        LOG.error(
+            "Instance " + instance + " in both ActivateInstances and DeactivateInstances list.");
+      }
+    }
+
+    // Add instances to current liveInstances
     ClusterState clusterState = new ClusterState();
     ConfigAccessor cfgAccessor = getConfigAccessor();
     HelixDataAccessor dataAccessor = getDataAccssor(clusterId);
     clusterState.resources = dataAccessor.getChildNames(dataAccessor.keyBuilder().idealStates());
     // Add existing live instances and new instances from user input to instances list.
-    clusterState.instances = dataAccessor.getChildNames(dataAccessor.keyBuilder().liveInstances());
-    clusterState.instances.addAll(inputFields.newInstances);
-
-    // Check if to be removed instances and old instances in swap node exist in live instance.
-    if (!inputFields.nodeSwap.isEmpty() || !inputFields.instancesToRemove.isEmpty()) {
-      Set<String> liveInstanceSet = new HashSet<>(clusterState.instances);
-      for (Map.Entry<String, String> nodeSwapPair : inputFields.nodeSwap.entrySet()) {
-        if (!liveInstanceSet.contains(nodeSwapPair.getKey())) {
-          throw new InvalidParameterException("Invalid input: instance [" + nodeSwapPair.getKey()
-              + "] in SwapInstances does not exist in cluster.");
-        }
-      }
-      for (String instanceToRemove : inputFields.instancesToRemove) {
-        if (!liveInstanceSet.contains(instanceToRemove)) {
-          throw new InvalidParameterException("Invalid input: instance [" + instanceToRemove
-              + "] in RemoveInstances does not exist in cluster.");
-        }
-      }
-      if (!inputFields.instancesToRemove.isEmpty()) {
-        clusterState.instances.removeIf(inputFields.instancesToRemove::contains);
-      }
-    }
+    Set<String> liveInstancesSet =
+        new HashSet<>(dataAccessor.getChildNames(dataAccessor.keyBuilder().liveInstances()));
+    liveInstancesSet.addAll(
+        inputFields.activatedInstances.stream().filter(s -> !liveInstancesSet.contains(s))
+            .collect(Collectors.toList()));
 
     // Read instance and cluster config.
-    // It will throw exception if there is no instanceConfig for newly added instance.
-    for (String instance : clusterState.instances) {
+    // Throw exception if there is no instanceConfig for activatedInstances instance.
+    for (String instance : liveInstancesSet) {
```

Comment:
1. if an instance is not alive, but it has an instance config, do we still add it to the input config list?
2. It seems that you will throw exception for the deactivated instances too if the corresponding config does not exist.

Reply:
Please correct me if I am wrong. When the code reach here, the `liveInstancesSet` contains current live instances and the instances in `activatedInstances` list? 
For instances in deactivated, these instances can only be in this list if they are currently under liveInstance, so they have corresponding config.

Reply:
Regarding the 2nd point, let's try to stick with the design. For deactivate, if the node does not exist at all, it is totally fine.

Reply:
In this implementation, it does not check if node exists for deactivated only check for activatedInstances.  
In the new commit, it reads all instance config (not only for live instances), and won't check if node exists for deactivated either. Only change the enabled state to disabled for to-be-deactivated instances.